README.md
# `resemble` Memory-Based Learning in Spectral Chemometrics
<!-- badges: start -->

[![Travis-CI Build Status](https://travis-ci.org/l-ramirez-lopez/resemble.svg?branch=master)](https://travis-ci.org/l-ramirez-lopez/resemble/)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/resemble)](https://CRAN.R-project.org/package=resemble)
[![Total_Downloads](http://cranlogs.r-pkg.org/badges/grand-total/resemble)](https://CRAN.R-project.org/package=resemble)

<img align="right" src="./man/figures/logo.png" width="15%">

<!-- badges: end -->

_Last update: 26.11.2021_

<em><p align="right"> Think Globally, Fit Locally (Saul and Roweis, 2003) </p></em>

## About
The `resemble` package provides high-performing functionality for 
data-driven modeling (including local modeling), nearest neighbor search and 
orthogonal projections in spectral data.

## Vignette
A new vignette for `resemble` explaining its core functionality is available 
at: https://cran.r-project.org/web/packages/resemble/vignettes/resemble.html

## Core functionality 

The core functionality of the package can be summarized into the following 
functions:
  
__`mbl`__: implements memory-based learning (MBL) for modeling and predicting 
continuous response variables. For example, it can be used to reproduce the 
famous LOCAL algorithm proposed by Shenk et al. (1997). In general, this 
function allows you to easily customize your own MBL regression-prediction 
method.

__`dissimilarity`__: Computes dissimilarity matrices based on various methods 
(e.g. Euclidean, Mahalanobis, cosine, correlation, moving correlation, 
Spectral information divergence, principal components dissimilarity and partial 
least squares dissimilarity).

__`ortho_projection`__: A function for dimensionality reduction using either 
principal component analysis or partial least squares (a.k.a projection to 
latent structures). 

__`search_neighbors`__: A function to efficiently retrieve from a reference set 
the k-nearest neighbors of another given data set.

## New version
During the recent lockdown we invested some of our free time to come up 
with a new version of our package. This new `resemble` 2.0 comes with MAJOR 
improvements and new functions! For these improvements major changes were 
required. The most evident changes are in the function and argument names.
These have been now adapted to properly follow the 
[tydiverse style guide](https://style.tidyverse.org/). A number of changes have 
been implemented for the sake of computational efficiency. These changes are 
documented in `inst\changes.md`.

New interesing functions and fucntionality are also available, for example, 
the `mbl()` function now allows sample spiking, where a 
set of reference observations can be forced to be included in the neighborhhoods 
of each sample to be predicted. The `serach_neighbors()` function efficiently 
retrieves from a refence set the k-nearest neighbors of another given data set. 
The `dissimilarity()` function computes dissimilarity matrices based on various 
metrics. 

## Installation 
If you want to install the package and try its functionality, it is very simple,
just type the following line in your `R` console:

```
install.packages('resemble')
```

If you do not have the following packages installed, it might be good to 
update/install them first
```
install.packages('Rcpp')
install.packages('RcppArmadillo')
install.packages('foreach')
install.packages('iterators')
```
__Note__: Apart from these packages we stronly recommend to download and install 
Rtools [https://cran.r-project.org/bin/windows/Rtools/](https://cran.r-project.org/bin/windows/Rtools/)). 
This is important for obtaining the proper C++ toolchain that might be needed 
for `resemble`.

Then, install `resemble`

You can also install the development version of `resemble` directly from github 
using [`devtools`](https://CRAN.R-project.org/package=devtools):
  
```
devtools::install_github("l-ramirez-lopez/resemble")
```

## Example

After installing `resemble` you should be also able to run the following lines:
  
```
library(resemble)
library(tidyr)
library(prospectr)
data(NIRsoil)

# Proprocess the data
NIRsoil <- NIRsoil[NIRsoil$CEC %>% complete.cases(),]
wavs <- as.numeric(colnames(NIRsoil$spc))

NIRsoil$spc_p <- NIRsoil$spc %>% 
  standardNormalVariate() %>% 
  resample(wavs, seq(min(wavs), max(wavs), by = 11)) %>% 
  savitzkyGolay(p = 1, w = 5, m = 1)

# split into calibration/training and test
train_x <- NIRsoil$spc_p[as.logical(NIRsoil$train), ]
train_y <- NIRsoil$CEC[as.logical(NIRsoil$train)]

test_x <- NIRsoil$spc_p[!as.logical(NIRsoil$train), ]
test_y <- NIRsoil$CEC[!as.logical(NIRsoil$train)]

# Use MBL as in Ramirez-Lopez et al. (2013)
sbl <- mbl(
  Xr = train_x, Yr = train_y, Xu = test_x,
  k = seq(50, 130, by = 20),
  method = local_fit_gpr(),
  control = mbl_control(validation_type = "NNv")
)
sbl
plot(sbl)
get_predictions(sbl)

````
<p align="center">
<img src="./man/figures/mbl.png" width="80%">
</p>
Figure 1. Standard plot of the results of the __`mbl`__ function.

[`resemble`](http://l-ramirez-lopez.github.io/resemble/) implements functions 
dedicated to non-linear modelling of complex visible and infrared spectral data 
based on memory-based learning (MBL, _a.k.a_ instance-based learning or local 
modelling in the chemometrics literature). The package also includes functions 
for: computing and evaluate spectral dissimilarity matrices, projecting the 
spectra onto low dimensional orthogonal variables, spectral neighbor search, etc. 

## Memory-based learning (MBL)

To expand a bit more the explanation on the `mbl` function, let's define 
first the basic input data:

* __Reference (training) set__: Dataset with *n* reference samples (e.g. spectral 
library) to be used in the calibration of spectral models. Xr represents the 
matrix of samples (containing the spectral predictor variables) and Yr represents 
a response variable corresponding to Xr.

* __Prediction set__ : Data set with _m_ samples where the response variable (Yu) 
is unknown. However it can be predicted by applying a spectral model 
(calibrated by using Xr and Yr) on the spectra of these samples (Xu). 

To predict each value in Yu, the `mbl` function takes each sample in Xu 
and searches in Xr for its _k_-nearest neighbours (most spectrally similar 
samples). Then a (local) model is calibrated with these (reference) neighbours 
and it immediately predicts the correspondent value in Yu from Xu. In the 
function, the _k_-nearest neighbour search is performed by computing spectral 
dissimilarity matrices between observations. The `mbl` function offers the 
following regression options for calibrating the (local) models:

__`'gpr'`__:          Gaussian process with linear kernel.

__`'pls'`__:          Partial least squares.     

__`'wapls'`__:        Weighted average partial least squares (Shenk et al., 1997).

Figure 2 illustrates the basic steps in MBL for a set of five observations.

<p align="center">
<img src="./vignettes/MBL.gif" width="50%">
</p>
Figure 2. Example of the main steps in memory-based learning for predicting a response variable in five different observations based on set of p-dimesnional variables.


## Citing the package
Simply type and you will get the info you need:
```
citation(package = "resemble")
```

## News
* 2020.04: [Tsakiridis et al. (2020)](https://www.sciencedirect.com/science/article/abs/pii/S0016706119308870?via%3Dihub), 
used the optmal principal components dissimilarity method implemented in `resemble` in combination with convolutional neural networks for simultaneous prediction of soil properties from vis-NIR spectra.

* 2019-04: [Tziolas et al. (2019)](https://www.sciencedirect.com/science/article/abs/pii/S0016706118307006), used
`resemble` to investigate on improved MBL methods for quantitative predictions 
of soil properties using NIR spectroscopy and geographical information. 

* 2019.03,08: [Tsakiridis et al. (2019a)](https://www.sciencedirect.com/science/article/abs/pii/S0925231220300291) and [Tsakiridis et al. (2019b)](https://www.sciencedirect.com/science/article/abs/pii/S1568494619302741),
compared several machine learning methods for predictive soil spectroscopy and 
show that MBL `resemble` offers highly competive results.

* 2020.01: [Sanderman et al., (2020)](https://acsess.onlinelibrary.wiley.com/doi/epdf/10.1002/saj2.20009) used `resemble` for the prediction of soil health indicatorsin the United States.

* 2019-03: Another paper using `resemble`... I published a [scientific paper](https://onlinelibrary.wiley.com/doi/10.1111/ejss.12752) were we used 
memory-based learning (MBL) for digital soil mapping. Here we use MBL to remove 
local calibration outliers rather than using this approach to overcome the typical 
complexity of large spectral datasets. (Ramirez‐Lopez, L., Wadoux, A. C.,
Franceschini, M. H. D., Terra, F. S., Marques, K. P. P., Sayão, V. M., & 
Demattê, J. A. M. (2019). Robust soil mapping at the farm scale with vis–NIR 
spectroscopy. European Journal of Soil Science. 70, 378–393).

* 2019-01: [In this scientific paper](https://www.mdpi.com/2571-8789/3/1/11/htm) 
we use `resemble` to model MIR spectra from a continental soil spectral library 
in United States. (Dangal, S.R., Sanderman, J., Wills, S. and Ramirez-Lopez, 
L., 2019. Accurate and Precise Prediction of Soil Properties from a Large
Mid-Infrared Spectral Library. Soil Systems, 3(1), p.11).

* 2019-03: [Jaconi et al. (2019)](https://www.sciencedirect.com/science/article/pii/S0016706118311340?via%3Dihub) implemented a memory-based learning algorithm (using `resemble`) to conduct 
accurate NIR predictions of soil texture at National scale in Germany. 
(Jaconi, A., Vos, C. and Don, A., 2019. Near infrared spectroscopy as an easy 
and precise method to estimate soil texture. Geoderma, 337, pp.906-913).

* 2018-12: [Chen, et al. (2018)](https://www.sciencedirect.com/science/article/pii/S0341816218304867) 
implemented a memory-based learning algorithm (using `resemble`) to improve the 
accuracy of NIR predictions of soil organic matter in China. 
(Hong, Y., Chen, S., Liu, Y., Zhang, Y., Yu, L., Chen, Y., Liu, Y., Cheng, H. 
and Liu, Y. 2019. Combination of fractional order derivative and memory-based learning 
algorithm to improve the estimation accuracy of soil organic matter by visible 
and near-infrared spectroscopy. Catena, 174, pp.104-116).

* 2018-11: In [this recent scientific paper](https://www.frontiersin.org/articles/10.3389/fpls.2018.01642/full) the authors used `resemble` to predict the chemoical composition of Common Beans in 
Spain. (Rivera, A., Plans, M., Sabaté, J., Casañas, F., Casals, J., Rull, A., & 
Simó, J. (2018). The Spanish core collection of common beans (Phaseolus 
vulgaris L.): an important source of variability for breeding chemical 
composition. Frontiers in Plant Science, 9).

* 2018-07: Another use-case of `resemble` is presented by [Gholizadeh et al.(2018)](https://www.mdpi.com/2072-4292/10/8/1172/htm) for a soil science 
application in Czech Republic. (Gholizadeh, A., Saberioon, M., Carmon, N., 
Boruvka, L. and Ben-Dor, E., 2018. Examining the Performance of PARACUDA-II 
Data-Mining Engine versus Selected Techniques to Model Soil Carbon from 
Reflectance Spectra. Remote Sensing, 10(8), p.1172).

* 2018-01: [Dotto, et al. (2018)](http://ufgrunwald.com/wp-content/uploads/2018/02/Dotto-et-al-2018.-Vis-NIR-Geoderma.pdf) have implemented memory-based learning with `resemble` to accurately 
predict soil organic Carbon at a region in Brazil. (Dotto, A. C., Dalmolin, 
R. S. D., ten Caten, A., & Grunwald, S. (2018). A systematic study on the 
application of scatter-corrective and spectral-derivative preprocessing for 
multivariate prediction of soil organic carbon by Vis-NIR spectra. Geoderma, 
314, 262-274).

* 2017-11: [Here](https://books.google.ch/books?hl=en&lr=&id=xlKODgAAQBAJ&oi=fnd&pg=PA129&dq=Evaluation+and+comparison+of+different+approaches+to+multi-product+brix+calibration+in+near-infrared+spectroscopy&ots=TGWzsRpP7X&sig=3u-pfuwkMZoxmhbfcx40eLiMOsY#v=onepage&q=Evaluation%20and%20comparison%20of%20different%20approaches%20to%20multi-product%20brix%20calibration%20in%20near-infrared%20spectroscopy&f=false) the authors predicted brix values in differet food products using memory-based learning implemented with `resemble`. 
(Kopf, M., Gruna, R., Längle, T. and Beyerer, J., 2017, March. Evaluation and 
comparison of different approaches to multi-product brix calibration in 
near-infrared spectroscopy. In OCM 2017-Optical Characterization of Materials-conference proceedings (p. 129). KIT Scientific Publishing).


NEWS
# News for the resemble package

## resemble 2.1 (piapia)
23.09.2021

### New features
- A argument named "seed" was added to the mbl function. It is used to gurantee 
reproducibility of cross-validation results. 

- A modified PLS method was implemented (see local_fit_pls and local_fit_wapls).
It uses correlation bteween response and predictors to derive the PLS weights. 

## Improvements and fixes
- A Bug in the computation of the explained variance of X for pls models was 
detected and fixed. The pls related functions were underestimating the amount 
of variance exmplained by each PLS component. The explained variance was being 
computed from the matrix of scores, in this version it is computed from the 
reconstructed spectra at each PLS iteration.

- Manual selection of components in pc_projection() and pls_projection() now 
accepts to select only 1 component (before the minimum was 2).

- pls_projection() now includes a new argument (method) to allow the user to select 
between the standard pls algorithm and a modified pls algorithm.

- ortho_diss(), dissimilarity() and search_neighbors() functions include a new 
dissimilarity method: "mpls" (modified pls).

- An internal function for stratified random sampling for cross validtaion 
purposes has been improved for computational speed. 

- The package was stripping some symbols for Rcpp functions in Makevars in order 
to reduce the installation size of the package. Now these lines have been 
commented to comply with CRAN policies.


## resemble 2.0 (gordillo)

* 02.07.2020 
During the recent lockdown we had the chance to inevest a enough time on the 
development of a new version of the package resemble. This new version comes 
with significant improvements as well as new functionality. For example, it now 
matches the tidyverse R style guide, it includes unit tests, includes new 
functionality for modeling, mbl is faster and a less memory-hungry.

### New features
- search_neighbors() function.
- dissimilarity() function.

## Improvements and fixes
- mbl is faster and a less memory-hungry.
- New vignette.
- unit tests have been introduced using the testthat package.

## Breaking changes

### orthoProjection, pcProjection, plsProjection (renamed to ortho_projection, 
pc_projection, pls_projection respectively):
- X2 argument renamed to Xu (for consistency throughout all the fucntions)
- Argument scaled renamed to .scale
- Argument max.iter renamed to max_iter
- Bug fix: when the "pca.nipals method was used and the method to select the pcs wa "opc",
  the function was returning 1 component less than the maximum requested.
- "pca.nipals" is now implemented in C++ via Rcpp
- Bug fix in plsProjection: when "cumvar" was used as the pcSelection method, an 
  error about data allocation in a matrix was retrieved
- Argument pcSelection renamed to pc_selection
- ... is deprecated in both pcProjection and plsProjection
- Argument cores is deprecated, it was used to set the number of cores in some 
c++ internal functions via OpenMP in Rcpp. 
- Names the following outputs have been changed: X.loadings, Y.loadings, sc.sdv 
and n.components. Their new names are: X_loadings, Y_loadings, sc_sdv and 
n_components.

### corDiss (renamed to cor_diss):
- X2 argument renamed to Xu (for consistency throughout all the fucntions)
- Argument scaled renamed to .scale
- default for .scale has changed from TRUE to FALSE
- the dimnames of the resulting matrix are now Xr_1... Xr_n (previusly Xr.1... Xr.n)

### fDiss (renamed to f_diss):
- X2 argument renamed to Xu (for consistency throughout all the fucntions)
- Argument scaled renamed to .scale
- default for .scale has changed from TRUE to FALSE
- the dimnames of the resulting matrix are now Xr_1... Xr_n (previusly Xr.1... Xr.n)
- argument method changed to diss_method

### sid:
- X2 argument renamed to Xu (for consistency throughout all the fucntions)
- Argument scaled renamed to .scale
- default for .scale has changed from TRUE to FALSE
- the dimnames of the resulting matrix are now Xr_1... Xr_n (previusly Xr.1... Xr.n)


### orthoDiss (renamed to ortho_diss):
- X2 argument renamed to Xu (for consistency throughout all the fucntions)
- Argument scaled renamed to .scale
- Argument local renamed to .local
- Argument pcSelection renamed to pc_selection
- Argument return.all renamed to compute_all
- Argument cores is deprecated, it wwas used to set the number of cores in some 
c++ internal functions via OpenMP in Rcpp. 
- When \code{.local = TRUE} a new output is produced: 'neighborhood_info' which 
is a data.frame containing the relevant information about the neighborhood of 
each sample (e.g. neighborhood indices, number of components used at each 
neighborhood, etc)
- Output global.variance.info has been renamed to global_variance_info


### simEval (renamed to sim_eval):
- argument sideInf renamed to side_info
- argument lower.tri renamed to lower_triangle
- argument cores renamed to omp_threads
- lower_triangle is deprecated. Now if a vector is passed to d, the function assumes
  that it is a lower triangle of a distance matrix
- Argument cores is deprecated, it wwas used to set the number of cores in some 
c++ internal functions via OpenMP in Rcpp.
  
  
### mbl
- pls.max.iter, pls.tol and noise.v were moved to mbl from mbl_control()
- Argument scaled (from mbl_control()) renamed to .scale and moved to mbl
- new arguments: gh and spike
- Argument pcSelection renamed to pc_selection
- Argument mblCtrl renamed to control
- Argument dissUsage renamed to diss_usage
- order of the Yr, Xr, Yu and Xu arguments has changed to Xr, Yr, Xu and Yu
- input type for the argument method has changed. Previously it received a 
character string  indicating the type of local regresion (i.e. "pls", 
"wapls1" or "gpr"). Now it receives an object of class local_fit which is output 
by the new local_fit fucntions. 
- dissimilarityM has been deprecated. It was used to pass a dissimilarity matrix 

code_ex.R
#-----------------------------------------------
# Load packages and prepare data
#-----------------------------------------------

# load packages
# Don't forget to install required packages if needed...
library(resemble);library(prospectr)
library(ggplot2)
library(foreach); library(reshape2)

# load dataset from the "Chimiométrie 2006" challenge
# see ?NIRsoil for an explanation of the variables
data(NIRsoil) 

# Filter the data using the Savitzky and Golay smoothing filter with a window size of
# 11 spectral variables and a polynomial order of 3 (no differentiation).
NIRsoil$spc_sg <- savitzkyGolay(X = NIRsoil$spc, p = 3, w = 11, m = 0)

# bin spectra every 10 nm 
NIRsoil$spc_sg <- binning(X=NIRsoil$spc, bin.size=5)
wavelength <- as.numeric(colnames(NIRsoil$spc)) # store wavelength position

# Select dependent variable
dep <- "Nt" # Total Nitrogen

# Remove NA values in dependent variable
anyna <- is.na(NIRsoil[,dep])
NIRsoil <- NIRsoil[!anyna,]

# Select cal & val dataset
Yr <- NIRsoil[as.logical(NIRsoil$train),dep] # dependent variable (calibration set)
Xr <- NIRsoil$spc_sg[as.logical(NIRsoil$train),] # spectral matrix (calibration set)
Yu <- NIRsoil[!as.logical(NIRsoil$train),dep] # dependent variable (validation set)
Xu <- NIRsoil$spc_sg[!as.logical(NIRsoil$train),] # spectra matrix (validation set)

#-----------------------------------------------
# Computing spectral dissimilarities
#-----------------------------------------------

# Compute distances
euDis <- fDiss(Xr = Xr,method="euclid") # euclidean distance
cosineDis <- fDiss(Xr = Xr, method="cosine") # cosine distance
corDis <- corDiss(Xr = Xr) # correlation distance
mcorDis <- corDiss(Xr = Xr, ws = 41) # moving correlation distance
pcaDis <- orthoDiss(Xr,pcSelection=list("cumvar",.99),method="pca",local=F)$dissimilarity # default parameter in orthoDiss
# Computation of a principal component dissimilarity matrix using the
# "opc" method for the selection of the principal components
opcDis <- orthoDiss(Xr,Yr = Yr,pcSelection = list("opc", 40),method = "pca")$dissimilarity
# Computation of a partial least squares (PLS) dissimilarity matrix 
plsDis <- orthoDiss(Xr,Yr = Yr,pcSelection = list("manual", 10),method = "pls")$dissimilarity

# Select one sample with the smallest mahalanobis distance to the centre of the data
# First Project in the PC space
pca <- prcomp(Xr,.center=T) # pca...
sc <- scale(pca$x,center=T,scale=T) # standardized score
pvar <- cumsum(pca$sdev^2/sum(pca$sdev^2)) # cumulative explained variance
pc <- max(which(pvar < .99)) + 1 # number of PC's accounting at least for .99 % of the variation
sc <- sc[,1:pc] # truncated score matrix
mahal <- mahalanobis(sc,center=colMeans(sc),cov=cov(sc)) # mahalanobis distance
id <- which.min(mahal) # index of the sample closest to the centre of the data

# Plot the closest sample to the mean
p <- ggplot(data=melt(Xr),aes(x=as.numeric(as.character(Var2)),y=value,group=Var1)) + geom_line(col="grey80",alpha=.5) + theme_bw() + labs(x="wavelength /nm",y="Absorbance") + scale_x_continuous(breaks=seq(1000,2500,250))
p + geom_line(data=melt(Xr[id,,drop=F]),col="red")

# Get distances to the selected sample and stack data
d <- as.data.frame(cbind(euDis[,id],cosineDis[,id],corDis[,id],mcorDis[,id],pcaDis[,id],opcDis[,id],plsDis[,id]))
colnames(d) <- paste0(c("euclidean","cosine","correlation","moving correlation","mahalanobis","opc","opc with pls")," distance")

# for each dissimilarity measure, find the 10 closest neighbours to the selected sample
knn <- sapply(d,function(x)order(x))[1:11,]
knn

# Plot
# merge spectra into one data.frame and format for ggplot2
spc <- foreach(i = 1:ncol(knn),.combine=rbind)%do%{
  data.frame(melt(Xr[knn[,i],]),d=colnames(knn)[i])
}
p <- ggplot(data=spc,aes(x=as.numeric(as.character(Var2)),y=value,group=Var1)) + geom_line(col="grey80") + facet_wrap(~d) + theme_bw() + labs(x="wavelength /nm",y="Absorbance") + scale_x_continuous(breaks=seq(1000,2500,250))
p + geom_line(data=melt(Xr[id,,drop=F]),col="red")

# Now we evaluate whether dissimarities are representative of the sample compositional variation
# using simEval
euSE <- simEval(d = euDis, sideInf = Yr)
cosineSE <- simEval(d = cosineDis, sideInf = Yr)
corSE <- simEval(d = corDis, sideInf = Yr)
mcorSE <- simEval(d = mcorDis, sideInf = Yr)
pcaSE <- simEval(d = pcaDis, sideInf = Yr)
opcSE <- simEval(d = opcDis, sideInf = Yr)
plsSE <- simEval(d = plsDis, sideInf = Yr)
sim <- list(euSE,cosineSE,corSE,mcorSE,pcaSE,opcSE,plsSE)
names(sim) <-  paste0(c("euclidean","cosine","correlation","moving correlation","mahalanobis","opc","opc with pls")," distance")

# stack rmsd and r (correlation) and plot
simplot <- (do.call(rbind,lapply(sim,function(x)x$eval)))
simplot$d <- factor(rownames(simplot),levels=rownames(simplot))
ggplot(data=melt(simplot),aes(y=value,x=d,fill=d)) + geom_bar(stat="identity") + facet_wrap(~variable,scale="free_y") + theme_bw() + labs(x="dissimilarities",y="") + scale_fill_brewer("Dissimilarities",palette="Set1") + scale_x_discrete(labels=abbreviate) + theme(legend.position="top")
# the 'opc' distance is the most representative of the soil compositional variation
# this is expected since 'opc' minimizes the distance between the closest neighbours and the training samples in the compositional space

#-----------------------------------------------
# Predict sample compostion with 'mbl'
# see ?mbl for other examples
#-----------------------------------------------

# Ex 1 : A simple Memory-based learning approach using partial least square regression and correlation distance
# First, set the parameters controlling the mbl function
ctrl <- mblControl(sm = "cor",              # dissimilarity
                   center = TRUE,           # Does the predictors need to be centered ?
                   scaled = TRUE,           # Does the predictors need to be scaled ?
                   valMethod = "none",      # No internal validation method
                   range.pred.lim = TRUE,   # Are the predictions constrained by the range of the response variable in each local model ?
                   progress = TRUE          # progress bar
                   )   
# Run mbl
mbl_cor <- mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu, mblCtrl = ctrl,
               dissUsage = "none",   # How the distance are used: 'none' = only in the selection of neighbours, 'weights' = distances used to computed sample weight through the tricubic function
               k = seq(25, 150, by = 25),  # a vector of number of neighbours to be tested
               method = "wapls1", # regression method: the prediction is a weighted average of values predicted with increasing pls components
               pls.c = c(4,15) # number of pls components                 
               )
mbl_cor
plot(mbl_cor) # plot of the error vs k

# Get predictions and plot pred-obs
preds <- getPredictions(mbl_cor)
predobs <- data.frame(pred=melt(preds),obs=Yu) 
predobs$k <- sub(".+_([0-9]+)","\\1",predobs$pred.variable)
p <- ggplot(data=predobs,aes(x=pred.value,y=obs))  + geom_point() + facet_wrap(~k) + theme_bw() + geom_abline(col="red") + labs(x="Predicted Nt",y="Observed Nt")
p 
p + geom_text(data=mbl_cor$YuPredictionStats,aes(x=-Inf,y=Inf,label=paste0("rmse = ",round(rmse,2))),vjust=2,hjust=-0.1) # this adds rmse info

# Ex 2 : same as  Ex. 1 but with the pc distance
ctrl$sm  <- "pc"
ctrl$pcSelection <- list(method="cumvar",value=.999)  # select as many PC's as to explain 99.9% of the spectral variation

mbl_pc <- mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu, mblCtrl = ctrl,
               dissUsage = "none",   
               k = seq(25, 150, by = 25), 
               method = "wapls1", 
               pls.c = c(4,15)
              )
mbl_pc

# Ex 3 : same as Ex. 2 but with the pls distance
ctrl$sm  <- "pls"
ctrl$pcSelection <- list(method="manual",value=10)

mbl_pls <- mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu, mblCtrl = ctrl,
              dissUsage = "none",   
              k = seq(25, 150, by = 25),  
              method = "wapls1", 
              pls.c = c(4,15)
              )
mbl_pls # this shows some improvement compared to pc distance...

# Ex 4 : same as Ex. 2 but with a user-defined distance matrix
# distances could be for instance computed on the 1st derivative spectra
Xr_der1 <- savitzkyGolay(NIRsoil$spc[as.logical(NIRsoil$train),], p = 5, w = 21, m = 1)
Xu_der1 <- savitzkyGolay(NIRsoil$spc[!as.logical(NIRsoil$train),], p = 5, w = 21, m = 1)
pls_dis <- orthoDiss(Xr=Xr_der1,X2 = Xu_der1, Yr = Yr,method = "pls", pcSelection = list("manual",10)) # pls distance in the 1st derivative space

ctrl$sm  <- "none" # we use pls_dis instead...
mbl_user_diss <- mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu,mblCtrl = ctrl,
                  dissimilarityM = pls_dis$dissimilarity, # user-specified dissimilarity matrix
                  dissUsage = "none",   
                  k = seq(50, 150, by = 25),  
                  method = "wapls1",  # we use the 'standard' pls algorithm...
                  pls.c = c(4,15)       # with a max of 15 components
                  )
mbl_user_diss

# Ex 5 : Let's try with another multivariate regression technique: pls
# the number of pls component for each local model is computed by cross-validation
ctrl$sm  <- "pls" 
mbl_pls_cv <- mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu, mblCtrl = ctrl,
               dissUsage = "none",   
               k = seq(25, 150, by = 25),  
               method = "pls",  # we use the 'standard' pls algorithm...
               pls.c = 15       # with a max of 15 components
               )
mbl_pls_cv # this shows some improvement compared to wapls1 regression

# Ex 6 : Let's try with another multivariate regression technique: gpr
mbl_gpr <- mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu, mblCtrl = ctrl,
               dissUsage = "none",   
               k = seq(25, 150, by = 25),  
               method = "gpr"  # predictions are done through gaussian process                   
                )
mbl_gpr # this is comparable with pls, without the need for tuning the number of components...

# Ex 7: A Memory-based learning approach (the spectrum-based learner), as implemeted in Ramirez-Lopez et al. (2013) 
sbl <- mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu,
           mblCtrl =  mblControl(valMethod = "none"), # most default values in mblControl correspond to the spectrum-based learner..., 
           dissUsage = "predictors", # the distance matrix is used as additional predictor
           k = seq(25, 150, by = 25), 
           method = "gpr" # predictions are done through gaussian process 
           )
sbl

#Plot all the results together
mymodels <- list(cor = mbl_cor,pc = mbl_pc, pls = mbl_pls, user_def = mbl_user_diss , pls_cv = mbl_pls_cv, gpr = mbl_gpr,sbl = sbl) 
tmp <- do.call(rbind,lapply(names(mymodels), function(x) data.frame(name = x, mymodels[[x]]$YuPredictionStats))) # stack model results
ggplot(data=tmp,aes(x=k,y=rmse,colour=name,linetype=name)) + geom_point() + geom_line() + theme_bw() +  ylim(c(0.4,0.8)) + scale_x_continuous(breaks=seq(25, 150, by = 25))

# Ex 7 : 
# In the previous examples, the selected samples for the local models (ie neighbours) are selected with a fixed number (k)
# but the neighborhood could be defined by distance tresholds
# this is implemented in mbl with the k.diss argument
mbl_kdiss <- mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu, 
               mblCtrl = mblControl(sm = "cor",valMethod = "none"), # correlation distance
               dissUsage = "none",   
               k.diss = c(0.25,0.1,0.05),  # samples with a correlation less than 0.75, 0.9, 0.95 with the sample to predict will not be included in the models
               k.range = c(20,nrow(Xr)), # the minimum and maximum number of samples to be included in the local calibrations
               method = "gpr"  
              )
# let's see what is the number of samples used in the calibrating models
res <-mbl_kdiss$results[[2]] # prediction table for correlation distance = 0.1
hist(as.numeric(as.factor(res$k)),xlab="k",main="")

# setting the tresholds might be less intuitive for other distances
# we suggest to pre-compute the dissimilarity matrix
# and look at the histogram of the dissimilarity values
# here is an example with opc distance
opc_dis <- orthoDiss(Xr=Xr,X2 = Xu, Yr = Yr,method = "pca", pcSelection = list("opc",40)) # pls distance in the 1st derivative space
ctrl$sm  <- "none" 
mbl_kdiss <- mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu, mblCtrl = ctrl,
                 dissimilarityM = opc_dis$dissimilarity,                  
                 dissUsage = "none",   
                 k.diss = c(0.05,0.1,0.2,0.5),  # samples with a pc distance > 0.1, 0.2, 0.5  with the sample to predict will not be included in the models
                 k.range = c(20,nrow(Xr)), # the minimum and maximum number of samples to be included in the local calibrations
                 method = "gpr"  
                )

# Ex 8: Internal validation
# resemble provides a way to make an internal validation of the results: 
# 'NNv': the nearest neighbour to the sample to predict is left out of the list of k nearest neighbours and predicted and compared with the actual value to compute a rmse
# 'loc_crossval' : each local model is cross-validated and the mean of the cross-validation results is computed
ctrl <- mblControl(sm="pls",pcSelection = list("manual",10),
                   valMethod = c("loc_crossval","NNv"),
                   resampling = 10, # controls the local cross-validation: this is the number of partition,
                   p = 0.25 # percentage of samples included in each partition
                  )
mbl_valid <- mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu, mblCtrl = ctrl,
               dissUsage = "none",   
               k = seq(25, 150, by = 25),  
               method = "gpr"  
              ) # note that the algorithm is slower when using internal validation

plot(mbl_valid) # shows the external (against Yr) and internal rmse, which seems a bit optimistic, compared to external validation

# Note that to be correct, external validation should be done with 3 sets:
# a reference (Xr), tuning and test sets
# the tuning set is used to find the best combination of MBL parameters (k, sm, etc..)
# which is then applied to predict the test set

#-----------------------------------------------
# Example of parallel execution
#-----------------------------------------------
# To speed-up mbl, the local models can be run in parallel
# using the 'parallel' package and the parallel back-end for the foreach %dopar% function 
# provided by 'doParallel'
library(parallel)
library(doParallel)
# here is the time spent with one core
ctrl <- mblControl(sm="pls",pcSelection = list("manual",10),progress = T)
system.time(mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu, mblCtrl = ctrl, dissUsage = "none", k = seq(25, 150, by = 25),  method = "gpr"))
# now with 4 cores
cl <- makeCluster(4) # create a set 4 running R copies
registerDoParallel(cl) # register them to work with foreach
system.time(mbl(Yr = Yr, Xr = Xr, Yu = Yu, Xu = Xu, mblCtrl = ctrl, dissUsage = "none", k = seq(25, 150, by = 25),  method = "gpr"))
# I get a ~3 x speed-up ...
registerDoSEQ() # un-register
stopCluster(cl) # delete R instances

# another example of parallel execution available in resemble is with the opc method
# opc can be time-intensive because it requires to compute a distance matrix and neareast-neighbours for each pc component
# In orthoProjection, the 'core' argument allows to do that in parallel 
#  (this will work only on Windows and Linux)
# Let's test this on a large matrix 
X <- matrix(rnorm(2*10^6),nrow=20000, ncol=100) # input (spectral) matrix of 20000 rows and 100 columns
y <- rnorm(2*10^4) # 'side' information used in the simEval function
system.time(orthoProjection(Xr = X,Yr=y, method = "pca",pcSelection = list("opc",40))) # this takes 285 sec on my machine
system.time(orthoProjection(Xr = X,Yr=y, method = "pca",pcSelection = list("opc",40),cores=4)) # and 136 sec with 4 cores...